Integrate Pictures backend with Posts

thecatfour opened the pull request:
### Category
- [X] **Feature**
- [ ] **Bug Fix**
- [X] **Testing**
- [ ] **Documentation**
- [ ] **Other**

### Domain
- [ ] **Frontend**
- [ ] **Backend**
- [X] **Both**

### Description
This PR adds image functionality for posts in the frontend and backend by updating post serializers to include associated images.

### Related Issue(s)
This PR fixes #77.

### New Additions

- **InputImage.tsx:** Repeatable component for inputting an image file into a form.
- **Pictures/signals.py:** Signals that delete unused images.

### Modifications
- Pictures/serializers.py: New serializers for adding post images.
- Pictures/views.py: New views for adding post images.
- Posts/serializers.py: New field in the post representation.
- forms/create-post-form.tsx: Now includes an image field and uses a form data object.
- forms/edit-post-form.tsx: Now includes an image field and uses a form data object.
- post-summary.tsx: Reformatted and includes a smaller version of the image as a thumbnail.
- post-summary-detailed.tsx: Now includes the image.
- url-segments.tx: Added new endpoints for pictures.

### New Dependencies
None.

### Visuals
This is what the post list and post summary components look like with the images:
![image](https://github.com/user-attachments/assets/52cac077-37a2-4a44-9455-e7f1cd0e8359)

This is what the detailed post component will look like with the images:
![image](https://github.com/user-attachments/assets/7753f1da-24ee-462b-b27a-7ab6659ffb87)

This is what using the image input feature looks like:
![image demonstration](https://github.com/user-attachments/assets/0831b35c-52c3-4981-8c96-83abbbbe7c07)

Posts can still be created without an image, and they will show the text bubble image in their post summary.

### Architecture of Your Changes
This PR updates the Pictures and Posts model in the backend by adding views, signals, and serializers. The signals are used to help remove images from the media folder that do not have a Picture object. Posts were updated to include a pictures list in the serializer. The serializer formats the returned image list in the format [ [picture id, picture url] ].

Posts that do not have an image will instead use a placeholder image that displays a text bubble. This image is uploaded on the frontend.

The frontend forms for posts were changed by updating how data is submitted. Now, the forms use a form data object to send the file to the backend. The forms also include an input component that displays a preview of the image that is being uploaded.

### Usage and Integration
Since the serializer sends a list of urls, it is possible to display any of the desired images. To display an image, do:
```
<img src={displayPicture(post.pictures[0][1])} />
```
Images are already displayed in post summary and detailed post summary, but they only display the first image currently. The serializer formats the returned image list in the format [ [picture id, picture url] ].

### Testing
Four tests were added to Pictures/tests.py to test functionality with posts. The tests check the following functions with posts:

- Adding a picture to a post
- Changing a picture on a post
- Deleting a picture from a post
- Getting a picture from the post serializer

![image](https://github.com/user-attachments/assets/b8743e11-b8e7-4042-8f90-b48fda8b9364)

### Mentions
@nnguyen144 and @imawful both use the post list component, so they may want to see the changes to posts. @leed85 may want to make sure Pictures will still be useable for profile.

thecatfour committed:
Update .gitignore

Now ignores the media file, which would store the images for posts.

thecatfour committed:
Update apps.py

Now includes the signal for when a post is deleted to also delete the image associated with the post.

thecatfour committed:
Create 0006_post_image.py

Added an image field for posts.

thecatfour committed:
Update models.py

Added an image field for posts.

thecatfour committed:
Added django-resized

Added django-resized to help store images in the backend.

thecatfour committed:
Create No Post Image.png

Thumbnail for posts that do not have an image.

thecatfour committed:
Changed files to accommodate images

Serializers and views needed to be updated to include the image field.

thecatfour committed:
Create signals.py

Created signal to delete an image if a post is deleted. Also deletes the linked image if the post is deleted through the admin panel.

thecatfour committed:
Update urls.py

Added urls to the media root.

thecatfour committed:
CSS changes

Created ImageInput CSS file to format the ImageInput component.

Most other changes were to make post summaries more compact.

thecatfour committed:
Update edit-post-form.tsx

Modified the edit post form to use a FormData object to pass the inputted information to the backend. This allows an image to be sent to the backend.

thecatfour committed:
Create ImageInput.tsx

Created a new component to help in uploading images. It displays a preview image as well as a button to help the user select a file to upload.

thecatfour committed:
Update post-summary-detailed.tsx

Post summary detailed object now includes the post's image if there is an image.

thecatfour committed:
Update post-summary.tsx

Post summary was redesigned to be more compact and include a small version of the post image as a thumbnail.

thecatfour committed:
Update definitions.ts

Added image field to the Post object.

thecatfour committed:
Added image upload field to post create

Post creation now uses a form data object to send the post fields, which allows the sending of an image for the post.

thecatfour committed:
Adjustments to formatting

Changed CSS files and how images are displayed in post summary detailed.

thecatfour committed:
Update tests.py

Added tests to check if media files are properly deleted and created when users upload images.

Tests use context managers with temporary directories and files to ensure the media directory is not changed.

thecatfour committed:
Posts standardization

Small changes that are meant to change where images are deleted. Now, deletion from the media folder is mostly handled through signals.py.

thecatfour committed:
Merge branch 'master' of https://github.com/UNLV-CS472-672/2024-F-GROUP3-RebelHub into posts-image

thecatfour committed:
Connect signals to Pictures

Connected signals.py to the Pictures model.

thecatfour committed:
Created Post Picture serializer

Added a serializer to create a picture object associated with a post.

thecatfour committed:
Create Picture signals

Created signals for pictures that delete images from the media file after their picture model is deleted or changed.

thecatfour committed:
Updated AddPictureToPostView

Modified the add picture view to use the serializer.

thecatfour committed:
Removed image testing stuff

Removed files and lines that I used to test image integration with the backend.

thecatfour committed:
Update post serializers to include image urls

Post serializers now includes the picture urls.

thecatfour committed:
Update views.py

thecatfour committed:
Update url-segments.ts

Added some picture related urls.

thecatfour committed:
Update definitions.ts

Added the picture urls list to the post type.

thecatfour committed:
Update create-post-form.tsx

Create-post-form now takes an image and does a separate API call to make the picture object.

thecatfour committed:
Post components use new pictures url

Post components were modified to display the images in the pictures[] list.

thecatfour committed:
Modified files to pass CI

thecatfour committed:
Add edit picture serializer

Added a serializer to help change an image for an existing picture object.

thecatfour committed:
Added tests for post pictures

Added two tests to check if pictures are correctly made when the picture views are called.

thecatfour committed:
Update urls.py

Added edit and delete post picture views.

thecatfour committed:
Update views.py

Added a edit post picture and delete post picture view.

thecatfour committed:
Update post serializers

Posts will now return a list of pictures associated with the post. If there are no pictures, then it will be an empty list.

thecatfour committed:
CSS changes

CSS changes to standardize where buttons for posts and comments are taken from. Also updates how the image is shown.

thecatfour committed:
Component CSS changes

Small CSS changes to help standardize buttons.

thecatfour committed:
Update edit-post-form.tsx

Edit post form can now create, change, and delete pictures associated with the post.

thecatfour committed:
Update ImageInput.tsx

Now includes a delete image button that removes the image from the post.

thecatfour committed:
Change how post.pictures works

Post.pictures is now a list of lists where each list contains the picture's id and the picture's url.

thecatfour committed:
Update url-segments.ts

Added picture endpoints.

thecatfour committed:
Added tests to Pictures

Added 4 unit tests to the Pictures model, one of which tests the Posts serializer.

thecatfour committed:
Merge branch 'master' of https://github.com/UNLV-CS472-672/2024-F-GROUP3-RebelHub into posts-image

thecatfour committed:
Update urls.py

Add picture and profile urls.

barrioskevin commented:
Hello, this pull request packs a lot of useful features in regards to the current Picture setup.

I appreciate the detail taken to add in signal to keep the media folder clean.

I have been testing out different images and I noticed that there is no scaling down of images, this could just be a nitpick but i uploaded some pretty large images and it kind of covers the entire post page. They look nice inside your modal but maybe we can at least scale how its viewed in the post so it's not too big. 

Other than that I haven't really found any issues. The test cases you wrote look good idk how you managed to come up with that but they pass. After this is merged I will attempt to add in the pictures for a hub because I failed my previous attempt. 

Thank you for your time,
- Kevin Barrios

thecatfour committed:
Modified detailed post summary

Post summary detailed component now scales the image based on the viewport. Its formatting was also changed to include the author's account button and a link to the hub that the post was created in.

thecatfour commented:
I slightly modified how detailed posts look, and I changed the scale of the images in the detailed post by giving it a max-height of 70vh. Here is an image of the changes:

![image](https://github.com/user-attachments/assets/2a6b1012-1e19-4867-80a8-f65a0b23e070)

thecatfour committed:
Update post-summary-detailed.tsx

Fixed a bug that caused the "in" word in the metadata to space incorrectly.

thecatfour committed:
Update requirements.txt

thecatfour committed:
Update models.py

barrioskevin commented:
As far as functionality goes I find no issues running your code.

The sizing of the picture looks much better when uploading large images. 

I have a few nitpicks but it's just about the styling nothing about the functionality. I think the overall style of our app should be updated soon so everything is consistent. I won't mention them here but It was just how the message is displayed under the photo when maybe we should  place the actual message above it. and then have a little bar holding all the images and we can click on them to see a full view. Similar to twitter. But again this is not worth changing in this pr.
